[errors] Add distinct error codes for docker not running

tejal29 opened the pull request:
Relates to #4692

This PR, tries to reduce BUILD_UNKNOWN for `docker` builder type when docker is not running. 

Description of changes:
1. Add a regular expression to recognize error thrown by docker when docker socket on host is not found. 
2. The way `BuildState` is represent right now, there is just one or most recent error code stored in `handler.state.BuildState.StatusCode`. From the event log, you can see the actual build error happens in the first build. The second build is cancelled as first build is failed. The `DevLoopEventFailed` uses this build state to propagate the error.
Ignoring cancellation errors, also make sure the build state has the error code with last legit build failure.


Before this PR, on master 

1. Stop your docker daemon. 
2. Try `skaffold dev`
With minikube deploy context skaffold fails when creating runner. 
```
➜  microservices git:(get_docker_json_output) ✗ skaffold dev             
creating runner: creating builder: getting docker client: getting minikube env: running [/Users/tejaldesai/Downloads/google-cloud-sdk2/bin/minikube docker-env --shell none -p minikube]
 - stdout: "\n\n"
 - stderr: "! Executing \"docker container inspect minikube --format={{.State.Status}}\" took an unusually long time: 7.36540945s\n* Restarting the docker service may improve performance.\nX Exiting due to GUEST_STATUS: state: unknown state \"minikube\": docker container inspect minikube --format=: exit status 1\nstdout:\n\n\nstderr:\nCannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?\n\n* \n* If the above advice does not help, please let us know: \n  - https://github.com/kubernetes/minikube/issues/new/choose\n"
 - cause: exit status 80
```
with GKE deploy context 
```
➜  microservices git:(get_docker_json_output) ✗ skaffold dev     
Listing files to watch...
 - leeroy-web
 - leeroy-app
Generating tags...
 - leeroy-web -> leeroy-web:v1.15.0-16-g8992c54bd
 - leeroy-app -> leeroy-app:v1.15.0-16-g8992c54bd
Checking cache...
 - leeroy-web: Not found. Building
 - leeroy-app: Not found. Building
Building [leeroy-web]...
exiting dev mode because first build failed: couldn't build "leeroy-web": docker build: Cannot connect to the Docker daemon at tcp://127.0.0.1:32770. Is the docker daemon running?
➜  microservices git:(get_docker_json_output) ✗ 
```


With this PR
Minikube Deploy context 
```
➜  microservices git:(get_docker_json_output) ✗ ../../out/skaffold dev             
Build Failed. Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Please check if docker is running.
➜  microservices git:(get_docker_json_output) ✗ 
```

GKE deploy context

```
➜  microservices git:(get_docker_json_output) ../../out/skaffold dev 
Listing files to watch...
 - leeroy-web
 - leeroy-app
Generating tags...
 - leeroy-web -> leeroy-web:v1.15.0-16-g8992c54bd
 - leeroy-app -> leeroy-app:v1.15.0-16-g8992c54bd
Checking cache...
 - leeroy-web: Not found. Building
 - leeroy-app: Not found. Building
Building [leeroy-web]...

Build Failed. Cannot connect to the Docker daemon at tcp://127.0.0.1:32770. Please check if docker is running.
➜  microservices git:(get_docker_json_output) 

```

Event API testing:

```
curl localhost:50052/v1/events              
{"result":{"timestamp":"2020-10-14T22:04:00.472529Z","event":{"metaEvent":{"entry":"Starting Skaffold: 
.....

{"result":{"timestamp":"2020-10-14T22:04:04.087004Z",
    "event":{"buildEvent":{"artifact":"leeroy-web",
                 "status":"In Progress"}},"entry":"Build started for artifact leeroy-web"}}


{"result":{"timestamp":"2020-10-14T22:04:08.037229Z",
    "event":{"buildEvent":{"artifact":"leeroy-web"," 
         status":"Failed",
        "err":"docker build: Cannot connect to the Docker daemon at tcp://127.0.0.1:32770. Is the docker daemon running?",
         "errCode":"BUILD_DOCKER_DAEMON_NOT_RUNNING",
             "actionableErr":{"errCode":"BUILD_DOCKER_DAEMON_NOT_RUNNING","message":"docker build: Cannot connect to the Docker daemon at tcp://127.0.0.1:32770. Is the docker daemon running?",
                "suggestions":[{"suggestionCode":"CHECK_DOCKER_RUNNING","action":"Please check if docker is running"}]}}},"entry":"Build failed for artifact leeroy-web"}}

{"result":{"timestamp":"2020-10-14T22:04:08.037291Z",
       "event":{"buildEvent":{"artifact":"leeroy-app",
         "status":"In Progress"}},"entry":"Build started for artifact leeroy-app"}}

{"result":{"timestamp":"2020-10-14T22:04:08.037789Z",
    "event":{"buildEvent":{"artifact":"leeroy-app",
      "status":"Failed",
      "err":"docker build: error during connect: Post \"https://127.0.0.1:32770/v1.24/build?buildargs=%7B%7D\u0026cachefrom=null\u0026cgroupparent=\u0026cpuperiod=0\u0026cpuquota=0\u0026cpusetcpus=\u0026cpusetmems=\u0026cpushares=0\u0026dockerfile=Dockerfile\u0026labels=null\u0026memory=0\u0026memswap=0\u0026networkmode=\u0026rm=0\u0026shmsize=0\u0026t=leeroy-app%3Av1.15.0-16-g8992c54bd\u0026target=\u0026ulimits=null\u0026version=\": context canceled","
      errCode":"BUILD_CANCELLED",
      "actionableErr":{"errCode":"BUILD_CANCELLED","message":"docker build: error during connect: Post \"https://127.0.0.1:32770/v1.24/build?buildargs=%7B%7D\u0026cachefrom=null\u0026cgroupparent=\u0026cpuperiod=0\u0026cpuquota=0\u0026cpusetcpus=\u0026cpusetmems=\u0026cpushares=0\u0026dockerfile=Dockerfile\u0026labels=null\u0026memory=0\u0026memswap=0\u0026networkmode=\u0026rm=0\u0026shmsize=0\u0026t=leeroy-app%3Av1.15.0-16-g8992c54bd\u0026target=\u0026ulimits=null\u0026version=\": context canceled"}}},"entry":"Build failed for artifact leeroy-app"}}

{"result":{"timestamp":"2020-10-14T22:04:08.042649Z",
      "event":{"devLoopEvent":{"status":"Failed",
      "err":{"errCode":"BUILD_DOCKER_DAEMON_NOT_RUNNING",
                  "message":"docker build: Cannot connect to the Docker daemon at tcp://127.0.0.1:32770. Is the docker daemon running?"}}},
             "entry":"Update failed with error code BUILD_DOCKER_DAEMON_NOT_RUNNING"}}
```

tejal29 commented:
Testing `GetState` in dev loop

1. Run `skaffold dev`
2. Pause Docker
3. Make a change in build dependencies. 
```
➜  microservices git:(get_docker_json_output) ✗ ../../out/skaffold dev --rpc-http-port=50052 --default-repo=gcr.io/tejal-test
Listing files to watch...
 - leeroy-web
 - leeroy-app
Generating tags...
 - leeroy-web -> gcr.io/tejal-test/leeroy-web:v1.15.0-16-g8992c54bd
 - leeroy-app -> gcr.io/tejal-test/leeroy-app:v1.15.0-16-g8992c54bd-dirty
Checking cache...
 - leeroy-web: Not found. Building
 - leeroy-app: Not found. Building
Building [leeroy-web]...
..
Building [leeroy-app]...
...
Starting deploy...
 - deployment.apps/leeroy-web created
 - service/leeroy-app created
 - deployment.apps/leeroy-app created
Deployments stabilized in 5.922908637s
Press Ctrl+C to exit

Watching for changes...
[leeroy-app] 2020/10/14 22:27:24 leeroy app server ready
[leeroy-web] 2020/10/14 22:27:25 leeroy web server ready
Generating tags...
 - leeroy-app -> gcr.io/tejal-test/leeroy-app:v1.15.0-16-g8992c54bd-dirty
Checking cache...
 - leeroy-app: Not found. Building
Building [leeroy-app]...

WARN[0059] Skipping deploy due to error: docker build: Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running? 
Watching for changes...
```

4. Hit the  get state API curl localhost:50052/v1/state 
```
➜  github-dashing git:(docker-support) ✗ curl localhost:50052/v1/state               
{"buildState":{"artifacts":{"leeroy-app":"Failed","leeroy-web":"Not Started"},"autoTrigger":true,"statusCode":"BUILD_DOCKER_DAEMON_NOT_RUNNING"},"deployState":{"status":"Not Started","autoTrigger":true},"statusCheckState":{"status":"Not Started"},"fileSyncState":{"status":"Not Started","autoTrigger":true},"metadata":{"build":{"numberOfArtifacts":2,"builders":[{"type":"DOCKER","count":2}],"type":"LOCAL"},"deploy":{"deployers":[{"type":"KUBECTL","count":1}],"cluster":"GKE"}}}%                                                         ➜  github-dashing git:(docker-support) ✗ 
```

5. Also check the event stream 
```

{"result":{"timestamp":"2020-10-14T22:27:56.538037Z","event":{"buildEvent":{"artifact":"leeroy-app","status":"Failed","err":"docker build: Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?","errCode":"BUILD_DOCKER_DAEMON_NOT_RUNNING","actionableErr":{"errCode":"BUILD_DOCKER_DAEMON_NOT_RUNNING","message":"docker build: Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?","suggestions":[{"suggestionCode":"CHECK_DOCKER_RUNNING","action":"Please check if docker is running"}]}}},"entry":"Build failed for artifact leeroy-app"}}

{"result":{"timestamp":"2020-10-14T22:27:56.538225Z",
   "event":{"devLoopEvent":{"status":"Failed",
    "err":{"errCode":"BUILD_DOCKER_DAEMON_NOT_RUNNING",
    "message":"docker build: Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?"}}},"entry":"Update failed with error code BUILD_DOCKER_DAEMON_NOT_RUNNING"}}

```

gsquared94 commented on pkg/skaffold/errors/err_map.go:
```diff
@@ -31,28 +31,59 @@ func re(s string) *regexp.Regexp {
 
 type problem struct {
 	regexp      *regexp.Regexp
-	description string
+	description func(error) string
```
Perhaps we're doing the same thing in two different ways: we match `regexp` against error to filter down to the `problem` but now we match error again to filter down to a description.

We can leave this as is and rather split `DockerConnectionFailedPrefix` into `DockerConnectionFailed` and `DockerConnectionFailedWithUnknownHost` for the case with known host address.
That'll keep `problem` struct at lowest granularity and easily referenceable. 

WDYT?

tejal29 replied:
The above approach still won't solve the problem. The reason why i added the `description` as a function was to extract the "address" where docker daemon should be running from the error message. 
e.g. if the error is 
```
exiting dev mode because first build failed: docker build: Cannot connect to the Docker daemon at tcp://127.0.0.1:32770. Is the docker daemon running?
```
I want to extract the "tcp://127.0.0.1:32770." and trim the error message as
```
Build failed. Cannot connect to the Docker daemon at tcp://127.0.0.1:32770. Is the docker daemon running?
```
When deploy context is not minikube, we could just return the error we receive as is.

When deploy context is minikube, the error we receive is
```
creating runner: creating builder: getting docker client: getting minikube env: running [/Users/tejaldesai/Downloads/google-cloud-sdk2/bin/minikube docker-env --shell none -p minikube]
 - stdout: "\n\n"
 - stderr: "! Executing \"docker container inspect minikube --format={{.State.Status}}\" took an unusually long time: 7.36540945s\n* Restarting the docker service may improve performance.\nX Exiting due to GUEST_STATUS: state: unknown state \"minikube\": docker container inspect minikube --format=: exit status 1\nstdout:\n\n\nstderr:\nCannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?\n\n* \n* If the above advice does not help, please let us know: \n  - https://github.com/kubernetes/minikube/issues/new/choose\n"
 - cause: exit status 80
```

which needs to be translated to 
```
Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
```

Another way to do this, if we use parse the "stdout" when we get this error from minikube and raise the error as "Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?"
maybe this is long term approach we want. 

Or, else we use a static string for both cases.
```
Cannot connect to the Docker daemon. Is the docker daemon running?
```

I hope it's clear.

In general, i feel we need a function to massage/trim the error we received to get rid of all the text added due to `errors.Wrap`. 

Or, we could add another field to  struct problem 
```
type problem struct {
	regexp      *regexp.Regexp
	description func(error) string
	errCode     proto.StatusCode
	suggestion  func(opts config.SkaffoldOptions) []*proto.Suggestion
        userMessage  func(error) string
}
```

gsquared94 commented on pkg/skaffold/errors/buildProblems.go:
```diff
@@ -23,6 +23,11 @@ import (
 	PushImageErrPrefix = "could not push image"
+	// Error Prefix matches error thrown by Docker
+	// See https://github.com/moby/moby/blob/master/client/errors.go#L18
+	DockerConnectionFailedPrefix = ".*Cannot connect to the Docker daemon.* "
```
`...Prefix` seems redundant on every variable and also isn't accurate since the pattern will match any substring and not just prefix.

If we want to group these under a common name how about:

```go
var errorPatterns = struct {
	PushImageErr string
}{
	PushImageErr: "could not push image",
}
```
which although verbose looks better when referenced:
```go
regexp:  re(fmt.Sprintf(".*%s.* denied: .*", errorPatterns.PushImageErr)),
```

tejal29 replied:
sounds good.

gsquared94 commented on pkg/skaffold/errors/errors.go:
```diff
@@ -67,11 +64,12 @@ func ShowAIError(err error) error {
 		if v.regexp.MatchString(err.Error()) {
 			if suggestions := v.suggestion(skaffoldOpts); suggestions != nil {
-				return fmt.Errorf("%s. %s", v.description, concatSuggestions(suggestions))
+				return fmt.Errorf("%s. %s", strings.Trim(v.description(err), "."), concatSuggestions(suggestions))
```
`strings.Trim(v.description(err), ".")`
can the `description` take care of the formatting and punctuation itself?

tejal29 replied:
the description is a func which returns a subset of error message from error. 
Moving it there, will force the developers to make sure, their regex trims "." which could be easily overlooked. 
Hence adding it here.

gsquared94 replied:
I meant to say if you did https://github.com/GoogleContainerTools/skaffold/pull/4914#discussion_r505405657 description could stay a string.

tejal29 replied:
i responded above. I don't want it to be just a string. i wanted to extract information from the error since 
1.  error is wrapped in many places. we cannot use `errors.Unwrap` to solve this as we are relying on wrapped text ;) 
2.  still wont solve the issue where deploy context is minikube.

gsquared94 commented on proto/skaffold.proto:
```diff
@@ -376,12 +376,15 @@ enum ClusterType {
     OK = 0;
 
-    // No Error codes
+    // Success Error codes
```
`// Success codes` instead of `// Success Error codes`?

tejal29 replied:
done.

gsquared94 commented on pkg/skaffold/event/event.go:
```diff
@@ -310,7 +310,7 @@ func BuildCanceled(imageName string) {
 	aiErr := sErrors.ActionableErr(sErrors.Build, err)
 	handler.stateLock.Lock()
-	handler.state.BuildState.StatusCode = aiErr.ErrCode
+	handler.state.BuildState.StatusCode = ignoreCancelled(handler.state.BuildState.StatusCode, aiErr.ErrCode)
```
`ignoreCancelled(handler.state.BuildState.StatusCode, aiErr.ErrCode)`

This makes me think that we're updating the `BuildState.StatusCode` in the wrong place since it seems to denote the status of the build sequence of several artifacts as a whole whereas `BuildFailed` is called for every artifact build failure.

Can we split out:
```go
// BuildFailed notifies that a build has failed.
func BuildFailed(imageName string, err error) {
	aiErr := sErrors.ActionableErr(sErrors.Build, err)
	handler.handleBuildEvent(&proto.BuildEvent{
		Artifact:      imageName,
		Status:        Failed,
		Err:           err.Error(),
		ErrCode:       aiErr.ErrCode,
		ActionableErr: aiErr})
}

// BuildSequenceFailed notifies that the build sequence has failed.
func BuildSequenceFailed(err error) {
	aiErr := sErrors.ActionableErr(sErrors.Build, err)
	handler.stateLock.Lock()
	handler.state.BuildState.StatusCode = aiErr.ErrCode
	handler.stateLock.Unlock()
}
```

and call `BuildSequenceFailed` in `scheduler.go` line 70
```go
if err := g.Wait(); err != nil {
	event.BuildSequenceFailed(err)
	return nil, err
}
```

tejal29 replied:
Yes. i think with the error group approach it would work.

gsquared94 commented on pkg/skaffold/event/event.go:
```diff
@@ -193,7 +193,7 @@ func DeployInProgress() {
 	aiErr := sErrors.ActionableErr(sErrors.Deploy, err)
 	handler.stateLock.Lock()
-	handler.state.DeployState.StatusCode = aiErr.ErrCode
+	handler.state.DeployState.StatusCode = ignoreCancelled(handler.state.DeployState.StatusCode, aiErr.ErrCode)
```
`event.DeployFailed` is raised only once in `runner/deploy.go`. Does this case actually happen for deploy too?

tejal29 replied:
we will working on adding deploy error codes soon as part of  #4849 and #4692

gsquared94 reviewed (changes requested):
PTAL at the comments. Looks good otherwise.

tejal29 commented:
Testing Notes:

As per @gsquared94 suggestion, moving updating build status code in `scheduler.go`, verified that devloop end event 

```

{"result":{"timestamp":"2020-10-15T18:23:24.696448Z","event":{"metaEvent":{"entry":"Starting Skaffold: \u0026{Version:v1.15.0-17-ga57104a62-dirty ConfigVersion:skaffold/v2beta9 GitVersion: GitCommit:a57104a625bd9bbe918546cb52c194a0ec423814 GitTreeState:dirty BuildDate:2020-10-15T10:25:16Z GoVersion:go1.14.3 Compiler:gc Platform:darwin/amd64}","metadata":{"build":{"numberOfArtifacts":2,"builders":[{"type":"DOCKER","count":2}],"type":"LOCAL"},"deploy":{"deployers":[{"type":"KUBECTL","count":1}],"cluster":"OTHER"}}}}}}
{"result":{"timestamp":"2020-10-15T18:23:24.696477Z","event":{"devLoopEvent":{"status":"In Progress"}},"entry":"Update initiated due to file change"}}
{"result":{"timestamp":"2020-10-15T18:23:27.273327Z","
   event":{"buildEvent":{"artifact":"leeroy-web","status":"In Progress"}},"entry":"Build started for artifact leeroy-web"}}

{"result":{"timestamp":"2020-10-15T18:23:55.946762Z","event":{
    "buildEvent":{"artifact":"leeroy-web","status":"Failed","err":"could not push image \"leeroy-web:v1.15.0-17-ga57104a62\": denied: requested access to the resource is denied",
   "errCode":"BUILD_PUSH_ACCESS_DENIED","actionableErr":{"errCode":"BUILD_PUSH_ACCESS_DENIED","message":"could not push image \"leeroy-web:v1.15.0-17-ga57104a62\": denied: requested access to the resource is denied","suggestions":[{"suggestionCode":"ADD_DEFAULT_REPO","action":"Trying running with `--default-repo` flag"}]}}},"entry":"Build failed for artifact leeroy-web"}}
{"result":{"timestamp":"2020-10-15T18:23:55.946806Z","event":{"buildEvent":{"artifact":"leeroy-app","status":"In Progress"}},"entry":"Build started for artifact leeroy-app"}}

{"result":{"timestamp":"2020-10-15T18:23:55.949108Z",
   "event":{"buildEvent":{"artifact":"leeroy-app","status":"Failed","err":"docker build: error during connect: Post \"http://%2Fvar%2Frun%2Fdocker.sock/v1.40/build?buildargs=%7B%7D\u0026cachefrom=null\u0026cgroupparent=\u0026cpuperiod=0\u0026cpuquota=0\u0026cpusetcpus=\u0026cpusetmems=\u0026cpushares=0\u0026dockerfile=Dockerfile\u0026labels=null\u0026memory=0\u0026memswap=0\u0026networkmode=\u0026rm=0\u0026shmsize=0\u0026t=leeroy-app%3Av1.15.0-17-ga57104a62\u0026target=\u0026ulimits=null\u0026version=\": context canceled",
   "errCode":"BUILD_CANCELLED","actionableErr":{"errCode":"BUILD_CANCELLED","message":"docker build: error during connect: Post \"http://%2Fvar%2Frun%2Fdocker.sock/v1.40/build?buildargs=%7B%7D\u0026cachefrom=null\u0026cgroupparent=\u0026cpuperiod=0\u0026cpuquota=0\u0026cpusetcpus=\u0026cpusetmems=\u0026cpushares=0\u0026dockerfile=Dockerfile\u0026labels=null\u0026memory=0\u0026memswap=0\u0026networkmode=\u0026rm=0\u0026shmsize=0\u0026t=leeroy-app%3Av1.15.0-17-ga57104a62\u0026target=\u0026ulimits=null\u0026version=\": context canceled"}}},"entry":"Build failed for artifact leeroy-app"}}

{"result":{"timestamp":"2020-10-15T18:23:55.951643Z","event":{
   "devLoopEvent":{"status":"Failed",
   "err":{"errCode":
     "BUILD_PUSH_ACCESS_DENIED","message":"could not push image \"leeroy-web:v1.15.0-17-ga57104a62\": denied: requested access to the resource is denied"}}},"entry":"Update failed with error code BUILD_PUSH_ACCESS_DENIED"}}
curl: (18) transfer closed with outstanding read data remaining

```

tejal29 commented:
TODO: 
Add test to make sure the error code is for the first failed artifact

tejal29 commented:
> PTAL at the comments. Looks good otherwise.

Thanks a lot @gsquared94 I addressed all your comment except for the `problem.description`. Let me know what you think!

gsquared94 reviewed (approved):
LGTM!

gsquared94 commented:
> TODO:
> Add test to make sure the error code is for the first failed artifact

I didn't see this TODO and merged on all green. Can you please add the tests in another PR? 
Thanks!

tejal29 commented:
will do. :)

tejal29 commented:
> > TODO:
> > Add test to make sure the error code is for the first failed artifact
> 
> I didn't see this TODO and merged on all green. Can you please add the tests in another PR?
> Thanks!

will do. I am thinking if it should be an integration test or just a unit test. Let me know if you have any thots.

gsquared94 commented:
> will do. I am thinking if it should be an integration test or just a unit test. Let me know if you have any thoughts or if you think you have already covered this case in `scheduler_test.go`

I only said it because you had a comment above, otherwise I don't have any thoughts about it 🙂.
I think it's covered in the last case in `TestInOrderForArgs` in `scheduler_test.go` but you can add a more explicit test if you want to.
